src/minidriver/minidriver.c

```diff
@@ -1384,12 +1523,35 @@ md_set_cmapfile(PCARD_DATA pCardData, struct md_file *file)
 	int ii, rv, conts_num, found_default = 0;
 	/* struct sc_pkcs15_data *data_object; */
 	struct sc_pkcs15_object *prkey_objs[MD_MAX_KEY_CONTAINERS];
+	pin_mode_t pin_mode = SCF_NONE;
+	int pin_cont_idx = -1;
 
 	if (!pCardData || !file)
 		return SCARD_E_INVALID_PARAMETER;
 
 	logprintf(pCardData, 2, "set 'cmapfile'\n");
 	vs = pCardData->pvVendorSpecific;
+	if (!vs)
+		return SCARD_E_INVALID_PARAMETER;
+
+	dwret = md_get_pin_by_role(pCardData, ROLE_USER, &vs->pin_objs[ROLE_USER]);
+	if (dwret != SCARD_S_SUCCESS)   {
+		logprintf(pCardData, 2, "Cannot get User PIN object");
+		return dwret;
```

Comment:
The old code (before these commits) also required presence of an user PIN - otherwise associate_card() would fail and so would then CardAcquireContext().

From a quick glance at the code it seems to me that it would be possible to make minidriver work even for cards that don't have an user PIN, however that would be a separate change from these
implementing multiple PINs support.